llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -12070,6 +12070,15 @@ SDValue DAGCombiner::visitVSELECT(SDNode *N) {
   if (SDValue F = extractBooleanFlip(N0, DAG, TLI, false))
     return DAG.getSelect(DL, VT, F, N2, N1);
 
+  // select (sext m), (add X, C), X --> (add X, (and C, (sext m))))
```

Comment:
I think we need to check `TLI.getBooleanContents(N0.getValueType()) == TargetLowering::ZeroOrNegativeOneBooleanContent`. Otherwise there's no guarantee that `m` is all zeros or all ones.

Reply:
@topperc  I will update it now, but can you please explain why we need this?

Reply:
Some targets you use 00000001 as a true value for a vector instead of 11111111. So we need to check what kind of target we're optimizing for.

Reply:
@topperc done